File: cpp/Mazes/searching.cpp
#include "searching.h"

namespace Boots
{
    template <typename T>
    const std::vector<T> GraphSearch<T>::dfs(const iGraph<T>& graph, const T& start)
    {
        std::vector<T> visited;
        std::vector<T> stack;
        stack.push_back(start);

        while (!stack.empty())
        {
            T node = stack.back();
            stack.pop_back();

            if (std::find(visited.begin(), visited.end(), node) == visited.end())
            {
                visited.push_back(node);
                for (const auto& neighbor : graph.getNeighbors(node))
                {
                    stack.push_back(neighbor);
                }
            }
        }

        return visited;
    }
    
    template <typename T>
    const std::vector<T> GraphSearch<T>::bfs(const iGraph<T>& graph, const T& start)
    {
        std::vector<T> visited;
        std::vector<T> queue;
        queue.push_back(start);

        while (!queue.empty())
        {
            T node = queue.front();
            queue.erase(queue.begin());

            if (std::find(visited.begin(), visited.end(), node) == visited.end())
            {
                visited.push_back(node);
                for (const auto& neighbor : graph.getNeighbors(node))
                {
                    queue.push_back(neighbor);
                }
            }
        }

        return visited;
    }

    template <typename T>
    const std::vector<T> GraphSearch<T>::dfsPath(const iGraph<T>& graph, const T& start, const T& end)
    {
        std::vector<T> path;
        std::vector<T> stack;
        stack.push_back(start);

        while (!stack.empty())
        {
            T node = stack.back();
            stack.pop_back();

            if (node == end)
            {
                path.push_back(node);
                return path;
            }

            if (std::find(path.begin(), path.end(), node) == path.end())
            {
                path.push_back(node);
                for (const auto& neighbor : graph.getNeighbors(node))
                {
                    stack.push_back(neighbor);
                }
            }
        }

        return {};
    }

    template <typename T>
    const std::vector<T> GraphSearch<T>::bfsPath(const iGraph<T>& graph, const T& start, const T& end)
    {
        std::vector<T> path;
        std::vector<T> queue;
        queue.push_back(start);

        while (!queue.empty())
        {
            T node = queue.front();
            queue.erase(queue.begin());

            if (node == end)
            {
                path.push_back(node);
                return path;
            }

            if (std::find(path.begin(), path.end(), node) == path.end())
            {
                path.push_back(node);
                for (const auto& neighbor : graph.getNeighbors(node))
                {
                    queue.push_back(neighbor);
                }
            }
        }

        return {};
    }

    template <typename T>
    const std::vector<T> GraphSearch<T>::dijkstra(const iWeightedGraph<T>& graph, const T& start, const T& end)
    {
        std::vector<T> path;
        std::vector<T> queue;
        queue.push_back(start);

        while (!queue.empty())
        {
            T node = queue.front();
            queue.erase(queue.begin());

            if (node == end)
            {
                path.push_back(node);
                return path;
            }

            if (std::find(path.begin(), path.end(), node) == path.end())
            {
                path.push_back(node);
                for (const auto& neighbor : graph.getNeighbors(node))
                {
                    queue.push_back(neighbor);
                }
            }
        }

        return {};
    }

    template <typename T>
    const std::vector<T> GraphSearch<T>::aStar(const iWeightedGraph<T>& graph, const T& start, const T& end, const std::function<double(const T&, const T&)>& heuristic)
    {
        std::vector<T> path;
        std::vector<T> queue;
        queue.push_back(start);

        while (!queue.empty())
        {
            T node = queue.front();
            queue.erase(queue.begin());

            if (node == end)
            {
                path.push_back(node);
                return path;
            }

            if (std::find(path.begin(), path.end(), node) == path.end())
            {
                path.push_back(node);
                for (const auto& neighbor : graph.getNeighbors(node))
                {
                    queue.push_back(neighbor);
                }
            }
        }

        return {};
    }

    template <typename T>
    const std::vector<T> GraphSearch<T>::bellmanFord(const iWeightedGraph<T>& graph, const T& start, const T& end)
    {
        std::vector<T> path;
        std::vector<T> queue;
        queue.push_back(start);

        while (!queue.empty())
        {
            T node = queue.front();
            queue.erase(queue.begin());

            if (node == end)
            {
                path.push_back(node);
                return path;
            }

            if (std::find(path.begin(), path.end(), node) == path.end())
            {
                path.push_back(node);
                for (const auto& neighbor : graph.getNeighbors(node))
                {
                    queue.push_back(neighbor);
                }
            }
        }

        return {};
    }

    template <typename T>
    const std::vector<T> GraphSearch<T>::floydWarshall(const iWeightedGraph<T>& graph, const T& start, const T& end)
    {
        std::vector<T> path;
        std::vector<T> queue;
        queue.push_back(start);

        while (!queue.empty())
        {
            T node = queue.front();
            queue.erase(queue.begin());

            if (node == end)
            {
                path.push_back(node);
                return path;
            }

            if (std::find(path.begin(), path.end(), node) == path.end())
            {
                path.push_back(node);
                for (const auto& neighbor : graph.getNeighbors(node))
                {
                    queue.push_back(neighbor);
                }
            }
        }

        return {};
    }
}

File: rust/searching.rs
#![deny(clippy::redundant_clone)]
#![deny(clippy::unwrap_used)]

use crate::graph::{Graph, Graphing, NeighborNode};
use crate::digraph::Digraph;
use crate::weighted_graph::{WeightedDigraph, WeightedGraph};
use std::hash::Hash;
pub trait SearchAlgorithm<T: Clone + Eq + Hash>: Graphing<T>
where
    <Self as Graphing<T>>::Neighbor: NeighborNode<T>,
{
    fn bfs(&self, start: &T, end: &T) -> Vec<T>{
        let mut path: Vec<T> = Vec::new();
        let mut stack: Vec<&T> = Vec::new();
        stack.push(start);

        while !stack.is_empty()
        {
            let node = stack.remove(0);

            if node == end
            {
                path.push(node.clone());
                return path;
            }

            if path.iter().find(|&x| *x == *node).is_none()
            {
                path.push(node.clone());
                if let Some(neighbors) = self.get_neighbors(node)
                {
                    for neighbor in neighbors {
                        let neigh_node: &T = NeighborNode::node(neighbor);
                        stack.push(neigh_node);
                    }
                }
            }
        }

        return Vec::new();
    }

    fn dfs(&self, start: &T, end: &T) -> Vec<T>{
        let mut path: Vec<T> = Vec::new();
        let mut stack: Vec<&T> = Vec::new();
        stack.push(start);

        while !stack.is_empty()
        {
            let node = stack.pop().unwrap();

            if node == end
            {
                path.push(node.clone());
                return path;
            }

            if path.iter().find(|&x| *x == *node).is_none()
            {
                path.push(node.clone());
                if let Some(neighbors) = self.get_neighbors(node)
                {
                    for neighbor in neighbors {
                        let neigh_node: &T = NeighborNode::node(neighbor);
                        stack.push(neigh_node);
                    }
                }
            }
        }

        return Vec::new();
    }
}

impl<T: Clone + Eq + Hash> SearchAlgorithm<T> for Graph<T>{}
impl<T: Clone + Eq + Hash> SearchAlgorithm<T> for Digraph<T>{}
impl<T: Clone + Eq + Hash> SearchAlgorithm<T> for WeightedGraph<T>{}
impl<T: Clone + Eq + Hash> SearchAlgorithm<T> for WeightedDigraph<T>{}

pub trait GreedySearch<T>
{
    fn dijkstra(&self, start: &T, end: &T) -> Vec<T>;
    fn bellman_ford(&self, start: &T, end: &T) -> Vec<T>;
    fn floyd_warshall(&self, start: &T, end: &T) -> Vec<T>;
    fn a_star(&self, start: &T, end: &T) -> Vec<T>;
}